File: challenges/Rho 2012/solution.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

#include <bits/stdc++.h>
using namespace std;

bool dfs(vector<int> &may, int w, int length) {
    if (may.size() > length) {
        return false;
    }
    if (may.back() == w) {
        return true;
    }
    if (may.size() == length) {
        return false;
    }
    for (int i = may.size() - 1; i >= 0; --i) {
        for (int j = i; j >= 0; --j) {
            const int x = may[i] + may[j];
            if (x > w) {
                continue;
            }
            if (x <= may.back() || x << (length - may.size() - 1) < w) {
                break;
            }
            may.push_back(x);
            if (dfs(may, w, length)) {
                return true;
            }
            may.pop_back();
        }
    }
    return false;
}

vector<int> solution(int A) {
    // Implement your solution here
    vector<int> r = {1};
    for (int length = 1; !dfs(r, A, length); ++length)
    ;
    return r;
}

File: challenges/Rho 2012/solution.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")


def dfs(may, w, length):
    if len(may) > length:
        return False
    if may[-1] == w:
        return True
    if len(may) == length:
        return False
    for i in range(len(may) - 1, -1, -1):
        for j in range(i, -1, -1):
            x = may[i] + may[j]
            if x > w:
                continue
            if x <= may[-1] or x << (length - len(may) - 1) < w:
                break
            may.append(x)
            if dfs(may, w, length):
                return True
            may.pop()
    return False

def solution(A):
    # Implement your solution here
    r = [1]
    length = 1
    while not dfs(r, A, length):
        length += 1
    return r
